### PR TITLE
Use Litenetlib for networking

### PR DESCRIPTION
**Pull request by feliwir**
Opening this to prevent any additional work going into the old networking backend

### CODE REVIEW

File: src/OpenSage.Game/Game.cs [hunk 1]
```diff
@@ -676,10 +676,10 @@ public AptWindow LoadAptWindow(string aptFileName)
                         players[i].SelectUnits(new[] { startingBuilding });
                     }
 
-                    if (players[i].IsHuman)
-                    {
-                        localPlayerIndex = i;
-                    }
+                    //if (players[i].IsHuman)
```
Comment (feliwir) [inline src/OpenSage.Game/Game.cs:679]: This would overwrite the localPlayerIndex passed to  `StartGame`. Not sure why it was in there

Comment (paavohuhtala) [reply src/OpenSage.Game/Game.cs:679]: Maybe the point was to autoselect the right player in single player maps?

Regardless, if this code is no longer deemed necessary it should be deleted.

File: src/OpenSage.Game/Game.cs [hunk 2]
```diff
@@ -853,6 +853,30 @@ private void LocalLogicTick(IEnumerable<InputMessage> messages)
             // We pass RenderTime to Scene2D so that the UI remains responsive even when the game is paused.
             Scene2D.LocalLogicTick(RenderTime, Scene3D?.LocalPlayer);
             Scene3D?.LocalLogicTick(MapTime, tickT);
+
+            // TODO: do this properly (this is a hack to call StartMultiplayerGame on the correct thread)
+            if (SkirmishManager.SkirmishGame?.ReadyToStart == true)
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Game.cs:858]: I would do just this:

```suggestion
            if (SkirmishManager.SkirmishGame?.ReadyToStart)
```

File: src/OpenSage.Game/Game.cs [hunk 3]
```diff
@@ -853,6 +853,30 @@ private void LocalLogicTick(IEnumerable<InputMessage> messages)
...
+            {
+                SkirmishManager.SkirmishGame.ReadyToStart = false;
+                StartMultiPlayerGame(
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Game.cs:861]: This code is very dense (pattern matching inside a constructor inside a LINQ expression inside an argument list) - I would personally split it into few more lines, at least by moving the player settings to a separate variable.

File: src/OpenSage.Game/Network/EchoConnection.cs [hunk 4]
```diff
@@ -1,42 +1,63 @@
...
 
-        public void Send(uint frame, List<Order> orders)
+        public virtual void Send(uint frame, List<Order> orders)
         {
-            _receivedPackets.Add(new ReceivedPacket
+            StorePacket(new SkirmishOrderPacket
             {
                 Frame = frame,
-                Orders = orders
+                Orders = orders.ToArray()
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Network/EchoConnection.cs:21]: This causes an array allocation (clone + copy) for every single sent packet. Probably not an issue at this point, but if we want to avoid frequent allocations as much as possible then we should figure out something to avoid this (later).

File: src/OpenSage.Game/Network/LobbyManager.cs [hunk 5]
```diff
@@ -1,114 +1,141 @@
...
+        private NetManager _manager;
+        private Thread _thread;
+        private bool _isRunning;
+        private NetPacketProcessor _processor;
+        private List<LobbyPlayer> _players = new List<LobbyPlayer>();
+
+        private static readonly NLog.Logger Logger = NLog.LogManager.GetCurrentClassLogger();
+
+        public IReadOnlyCollection<LobbyPlayer> Players => _players;
+        public string LocalAddress => NetUtils.GetLocalIp(LocalAddrType.IPv4);
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Network/LobbyManager.cs:26]: This is fine for now, but I think for some computers this will pick the wrong IP address if you have multiple network adapters, including virtual ones.

File: src/OpenSage.Game/Network/LobbyManager.cs [hunk 6]
```diff
@@ -1,114 +1,141 @@
...
-        public static IPAddress GetBroadcastAddress(IPAddress address, IPAddress mask)
+        private void LobbyBroadcastReceived(LobbyBroadcastPacket packet, IPEndPoint endPoint)
         {
-            uint ipAddress = BitConverter.ToUInt32(address.GetAddressBytes(), 0);
-            uint ipMaskV4 = BitConverter.ToUInt32(mask.GetAddressBytes(), 0);
-            uint broadCastIpAddress = ipAddress | ~ipMaskV4;
-
-            return new IPAddress(BitConverter.GetBytes(broadCastIpAddress));
-        }
+            // Logger.Trace($"Received {nameof(LobbyBroadcastPacket)} from { endPoint }.");
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Network/LobbyManager.cs:122]:
```suggestion
```

File: src/OpenSage.Game/Network/SkirmishManager.cs [hunk 7]
```diff
@@ -1,350 +1,255 @@
...
 
-                Serializer.Serialize(_serverConnection.GetStream(), selfPacket);
-            }
-        }
+            var server = _manager.Connect(endPoint.Address.ToString(), Ports.SkirmishHost, _writer);
 
-        public SkirmishManager(Game game)
-        {
-            _game = game;
+            //_writer.Put((byte) PacketType.SkirmishClientUpdate);
```
Comment (paavohuhtala) [inline src/OpenSage.Game/Network/SkirmishManager.cs:196]: Remove commented out code.

File: src/OpenSage.Mods.Generals/Gui/LanGameOptionsMenuCallbacks.cs [hunk 8]
```diff
@@ -48,32 +49,48 @@ public static void LanGameOptionsMenuInit(Window window, Game game)
...
+                var playerCombo = (ComboBox)window.Controls.FindControl($"LanGameOptionsMenu.wnd:ComboBoxPlayer{slot.Index}");
+                if (slot.State == Network.SkirmishSlotState.Human)
+                {
+                    playerCombo.Controls[0].Text = slot.PlayerName;
+                }
+                else
+                {
+                    playerCombo.Controls[0].Text = slot.State.ToString();
+                }
+                //playerCombo.SelectedIndex = (int) slot.;
```
Comment (paavohuhtala) [inline src/OpenSage.Mods.Generals/Gui/LanGameOptionsMenuCallbacks.cs:88]:
```suggestion
```

File: src/OpenSage.Mods.Generals/Gui/OptionsMenuCallbacks.cs [hunk 9]
```diff
@@ -28,7 +28,7 @@ public static void OptionsMenuInit(Window window, Game game)
         {
             var comboBoxIP = (ComboBox) window.Controls.FindControl(ComboBoxIPPrefix);
             var items = new List<ListBoxDataItem>();
-            items.Add(new ListBoxDataItem(game.LobbyManager.Unicast.Address, new[] { game.LobbyManager.Unicast.Address.ToString() }, comboBoxIP.TextColor));
+            // items.Add(new ListBoxDataItem(game.LobbyManager.Unicast.Address, new[] { game.LobbyManager.Unicast.Address.ToString() }, comboBoxIP.TextColor));
```
Comment (paavohuhtala) [inline src/OpenSage.Mods.Generals/Gui/OptionsMenuCallbacks.cs:31]:
```suggestion
```

File: src/OpenSage.Game/Game.cs (hunk 3 above)
Comment (tgjones) [reply src/OpenSage.Game/Game.cs:872]: I second what @paavohuhtala said.

File: src/OpenSage.Game/Graphics/Cameras/Camera.cs [hunk 10]
```diff
@@ -80,9 +80,11 @@ public float NearPlaneDistance
 
         public Camera(Func<Viewport> getViewport)
         {
+
             _getViewport = getViewport;
             _viewport = getViewport();
 
+            View = Matrix4x4.Identity;
```
Comment (tgjones) [inline src/OpenSage.Game/Graphics/Cameras/Camera.cs:87]: Why was this necessary?

**Comment by tgjones** [comment]
This is very exciting. Thank you!

File: src/OpenSage.Game/Game.cs (hunk 3 above)
Comment (feliwir) [reply src/OpenSage.Game/Game.cs:872]: @tgjones we've already split playerSettings into a seperate variable, but this probably still should be simpler

File: src/OpenSage.Game/Graphics/Cameras/Camera.cs (hunk 10 above)
Comment (feliwir) [reply src/OpenSage.Game/Graphics/Cameras/Camera.cs:87]: Sometimes the View matrix was not initialized resulting in crashes with NaN / division by zero. This gets initialized from outside during a tick, but apparently the order seems to be different for networking. However visually everything is still the same
